tests/reproducibility/test_graphsage.py

```diff
@@ -145,6 +145,67 @@ def gs_nai(
     return model
 
 
+def gs_link_pred_model(num_samples, generator, optimizer, bias, dropout, normalize):
+    layer_sizes = [50] * len(num_samples)
+    graphsage = GraphSAGE(
+        layer_sizes=layer_sizes,
+        generator=generator,
+        bias=bias,
+        dropout=dropout,
+        normalize=normalize,
+    )
+    # Build the model and expose input and output sockets of graphsage, for node pair inputs:
+    x_inp, x_out = graphsage.build()
+    pred = link_classification(
+        output_dim=1, output_act="relu", edge_embedding_method="ip"
+    )(x_out)
+
+    model = tf.keras.Model(inputs=x_inp, outputs=pred)
+
+    model.compile(
+        optimizer=optimizer, loss=tf.keras.losses.binary_crossentropy,
+    )
+
+    return model
+
+
+def gs_link_prediction(
+    g,
+    edge_ids,
+    edge_labels,
+    num_samples,
+    optimizer,
+    batch_size=4,
+    epochs=4,
+    bias=True,
+    dropout=0.0,
+    normalize="l2",
+    seed=0,
+    shuffle=True,
+):
+    set_seed(seed)
+    tf.random.set_seed(seed)
+    if shuffle:
+        random.seed(seed)
+
+    generator = GraphSAGELinkGenerator(g, batch_size, num_samples)
+    train_gen = generator.flow(edge_ids, edge_labels, shuffle=True)
+
+    model = gs_link_pred_model(
+        num_samples, generator, optimizer, bias, dropout, normalize
+    )
+
+    model.fit_generator(
```

Comment:
In light of #803, maybe we shouldn't add new code that uses `fit_generator`?

Reply:
Hmm actually I think it might be better to do the switch all at once and keep using `fit_generator` here. The deprecation of `fit_generator` only occurred with tensorflow 2.1, and I'm getting different behaviour from using `fit` with a generator depending on the tensorflow version - tensorflow 2.1 works fine, but 2.0 gives me errors, so I think we want to have `tensorflow>=2.1` then make the switch?

Reply:
Error for reference:
```

    model = func()
tests/reproducibility/test_graphsage.py:247: in <lambda>
    shuffle=shuffle,
tests/reproducibility/test_graphsage.py:204: in gs_link_prediction
    shuffle=shuffle,
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py:728: in fit
    use_multiprocessing=use_multiprocessing)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2.py:224: in fit
    distribution_strategy=strategy)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2.py:547: in _process_training_inputs
    use_multiprocessing=use_multiprocessing)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2.py:606: in _process_inputs
    use_multiprocessing=use_multiprocessing)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/data_adapter.py:613: in __init__
    output_shapes=nested_shape)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/data/ops/dataset_ops.py:540: in from_generator
    output_types, tensor_shape.as_shape, output_shapes)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/data/util/nest.py:471: in map_structure_up_to
    results = [func(*tensors) for tensors in zip(*all_flattened_up_to)]
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/data/util/nest.py:471: in <listcomp>
    results = [func(*tensors) for tensors in zip(*all_flattened_up_to)]
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/framework/tensor_shape.py:1216: in as_shape
    return TensorShape(shape)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/framework/tensor_shape.py:776: in __init__
    self._dims = [as_dimension(d) for d in dims_iter]
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/framework/tensor_shape.py:776: in <listcomp>
    self._dims = [as_dimension(d) for d in dims_iter]
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/framework/tensor_shape.py:718: in as_dimension
    return Dimension(value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'Dimension' object has no attribute '_value'") raised in repr()] Dimension object at 0x144d2df90>
value = (None, None, None)

    def __init__(self, value):
      """Creates a new Dimension with the given value."""
      if value is None:
        self._value = None
      elif isinstance(value, Dimension):
        self._value = value.value
      elif isinstance(value, dtypes.DType):
        raise TypeError("Cannot convert %s to Dimension" % value)
      else:
>       self._value = int(value)
E       TypeError: int() argument must be a string, a bytes-like object or a number, not 'tuple'

../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/framework/tensor_shape.py:193: TypeError
---------------------------------------- Captured stdout call -----------------------------------------
link_classification: using 'ip' method to combine node embeddings into edge embeddings
_____________________________________ test_link_prediction[False] _____________________________________

petersen_graph = <stellargraph.core.graph.StellarGraph object at 0x145114a90>, shuffle = False

    @pytest.mark.parametrize("shuffle", [True, False])
    def test_link_prediction(petersen_graph, shuffle):
        num_examples = 4
        edge_ids = np.random.choice(petersen_graph.nodes(), size=(num_examples, 2))
        edge_labels = np.random.choice([0, 1], size=num_examples)
        assert_reproducible(
>           lambda: gs_link_prediction(
                petersen_graph,
                edge_ids,
                edge_labels,
                [2, 2],
                tf.optimizers.Adam(1e-3),
                shuffle=shuffle,
            )
        )

tests/reproducibility/test_graphsage.py:241:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/reproducibility/fixtures.py:35: in assert_reproducible
    model = func()
tests/reproducibility/test_graphsage.py:247: in <lambda>
    shuffle=shuffle,
tests/reproducibility/test_graphsage.py:204: in gs_link_prediction
    shuffle=shuffle,
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py:728: in fit
    use_multiprocessing=use_multiprocessing)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2.py:224: in fit
    distribution_strategy=strategy)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2.py:547: in _process_training_inputs
    use_multiprocessing=use_multiprocessing)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/training_v2.py:606: in _process_inputs
    use_multiprocessing=use_multiprocessing)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/keras/engine/data_adapter.py:613: in __init__
    output_shapes=nested_shape)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/data/ops/dataset_ops.py:540: in from_generator
    output_types, tensor_shape.as_shape, output_shapes)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/data/util/nest.py:471: in map_structure_up_to
    results = [func(*tensors) for tensors in zip(*all_flattened_up_to)]
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/data/util/nest.py:471: in <listcomp>
    results = [func(*tensors) for tensors in zip(*all_flattened_up_to)]
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/framework/tensor_shape.py:1216: in as_shape
    return TensorShape(shape)
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/framework/tensor_shape.py:776: in __init__
    self._dims = [as_dimension(d) for d in dims_iter]
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/framework/tensor_shape.py:776: in <listcomp>
    self._dims = [as_dimension(d) for d in dims_iter]
../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/framework/tensor_shape.py:718: in as_dimension
    return Dimension(value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'Dimension' object has no attribute '_value'") raised in repr()] Dimension object at 0x1452ab5d0>
value = (None, None, None)

    def __init__(self, value):
      """Creates a new Dimension with the given value."""
      if value is None:
        self._value = None
      elif isinstance(value, Dimension):
        self._value = value.value
      elif isinstance(value, dtypes.DType):
        raise TypeError("Cannot convert %s to Dimension" % value)
      else:
>       self._value = int(value)
E       TypeError: int() argument must be a string, a bytes-like object or a number, not 'tuple'

../../../miniconda3/envs/stellargraph-dev/lib/python3.7/site-packages/tensorflow_core/python/framework/tensor_shape.py:193: TypeError
```
I'm guessing `fit` didn't support all the different ways the generator inputs can be formed until 2.1

Reply:
What a pest ☹️ ah well.